connectors/sources/gcs.py

```diff
@@ -0,0 +1,304 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""Google Cloud Storage source module responsible to fetch documents from Google Cloud Storage.
+"""
+import asyncio
+import json
+import os
+import urllib.parse
+from functools import partial
+
+from aiogoogle import Aiogoogle
+from aiogoogle.auth.creds import ServiceAccountCreds
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource
+
+CLOUD_STORAGE_READ_ONLY_SCOPE = "https://www.googleapis.com/auth/devstorage.read_only"
+CLOUD_STORAGE_BASE_URL = "https://console.cloud.google.com/storage/browser/_details/"
+API_NAME = "storage"
+API_VERSION = "v1"
+BLOB_ADAPTER = {
+    "_id": "id",
+    "component_count": "componentCount",
+    "content_encoding": "contentEncoding",
+    "content_language": "contentLanguage",
+    "created_at": "timeCreated",
+    "last_updated": "updated",
+    "metadata": "metadata",
+    "name": "name",
+    "size": "size",
+    "storage_class": "storageClass",
+    "_timestamp": "updated",
+    "type": "contentType",
+    "url": "selfLink",
+    "version": "generation",
+    "bucket_name": "bucket",
+}
+SUPPORTED_FILETYPE = [".py", ".rst", ".rb", ".sh", ".md", ".txt"]
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_WAIT_MULTIPLIER = 2
+DEFAULT_CONTENT_EXTRACTION = True
+DEFAULT_FILE_SIZE_LIMIT = 10485760
+STORAGE_EMULATOR_HOST = os.getenv(
+    key="STORAGE_EMULATOR_HOST", default=None
+)  # For end to end tests set this environment variable with target host.
+
+
+class GoogleCloudStorageDataSource(BaseDataSource):
+    """Class to fetch documents from Google Cloud Storage."""
+
+    def __init__(self, connector):
+        """Setup connection to the Google Cloud Storage Client.
+
+        Args:
+            connector (BYOConnector): Object of the BYOConnector class.
+        """
+        super().__init__(connector=connector)
+        self.credentials_path = self.configuration["credentials_path"]
+        self.service_account_credentials = ServiceAccountCreds(
+            scopes=[CLOUD_STORAGE_READ_ONLY_SCOPE],
+            **json.load(fp=open(file=self.credentials_path)),
+        )
+        self.user_project_id = self.service_account_credentials.project_id
+        self.retry_count = int(
+            self.configuration.get("retry_count", DEFAULT_RETRY_COUNT)
+        )
+        self.enable_content_extraction = self.configuration.get(
+            "enable_content_extraction", DEFAULT_CONTENT_EXTRACTION
+        )
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for Google Cloud Storage.
+
+        Returns:
+            dictionary: Default configuration.
+        """
+        return {
+            "credentials_path": {
```

Comment:
using a local file will not work on cloud. We should allow people to pass a string here, instead that contains the JSON file

Reply:
The connector now accepts the credentials instead of the file path. It assumes that the user will copy/paste the contents of the service-account credentials file.